smtc-modem-cores

lthiery opened the pull request:
After seeing [Semtech's new reference drivers](https://github.com/Lora-net/SWL2001), I thought it would be interesting to start using them in testing the lora-phy implementation. Full chip support exists in these reference drivers, so it would be further useful for developing sx128x or lr1110 implementations.

Furthermore, a no-async device implementation could be based on this bindgen package.

Per review feedback, the bindings crates live in their own repository: https://github.com/lora-rs/smtc-modem-cores. This PR is just the lora-phy test integration — the bindings come in as a git dev-dependency, and the tests compare the SPI byte stream our sx126x driver produces against the reference driver's, starting with `set_sleep` (cold and warm start).

On the C-callback wiring: an earlier iteration used a `concrete_export!` macro that made the user define the `#[no_mangle] sx126x_hal_write` symbol for their concrete SPI type, which was awkward and limited each binary to a single SPI type (duplicate symbol otherwise). The bindings crate now owns that symbol itself and dispatches through a monomorphized fn pointer stored at the head of the `repr(C)` context — `Context::new(spi)` is the whole user-facing API, and multiple SPI types can coexist in one binary. Extending to `sx126x_hal_read`/`reset`/`wakeup` is just more fields in that table.

**Update: lr1110 coverage.** The bindings repo now also builds SWL2001's lr11xx radio core (its HAL differs — two-transaction reads plus a direct read, so the vtable carries three entry points), and this branch adds 46 comparison tests for the lr1110 driver mirroring the sx126x suite: system commands, the LoRa TX/RX/CAD paths, TX power on all three PAs, regmem, GFSK configuration and LR-FHSS framing.

The comparison immediately caught real encoding bugs in the lr1110 driver, fixed here in their own commit: the SX126x sleep-config bit layout (every sleep was cold, losing retention), the SX126x mantissa/exponent symbol timeout (8 symbols became 32), a stray offset byte in WriteBuffer8 (would have been transmitted as the first payload byte), a zero length field in ReadBuffer8, SX126x-style bitrate/deviation conversions in GFSK modulation params (the lr11xx takes raw bps/Hz), a GFSK packet-status decode reading payload length from an RSSI byte, missing AutoTxRx fields, and the high-ACP workaround the reference applies on every SetTx/SetRx/SetCad.

**Update: sx127x coverage.** The register-based sx127x needed a different comparison approach: the two drivers legitimately factor their register traffic differently (SWL2001 burst-writes register runs and keeps shadow copies; ours does single-register read-modify-write over SPI), so the sx127x fixture is a small register file both drivers execute against, and equality is the chip-visible outcome — final register state plus the FIFO byte stream. 16 tests cover the LoRa config surface, TX/CAD flows, RX reads and decoded-value parity.

This round caught five bugs in our sx127x driver, fixed in their own commit: the PLL-step shortcut truncated the low 8 bits so fractional-MHz channels (868.1, 903.9 MHz) sat up to ~15 kHz off carrier; PA_BOOST requests above 17 dBm used the PaDac-off power formula and overflowed the 4-bit OutputPower field; RFO requests at or below 0 dBm cast a negative OutputPower to u8 and corrupted RegPaConfig; the RegDetectionOptimize RMW mask dropped the AutomaticIFOn bit; and packet RSSI skipped the 16/15 linearization in the negative-SNR branch. It also surfaced a reference-driver bug worth knowing about: SWL2001's `sx127x_set_ocp_value` masks the trim with the inverted-form `OCP_TRIM_MASK`, so it always programs 45 mA regardless of the value passed.

With sx126x, lr1110 and sx127x covered, every lora-phy driver with a SWL2001 counterpart now has a comparison suite.

plaes commented on .github/workflows/rust.yml:
```diff
@@ -19,11 +19,17 @@ jobs:
           access_token: ${{ github.token }}
 
+      - name: Setup apt packages
+        run: sudo apt-get update && sudo apt-get install -y cmake llvm-dev clang libclang-dev
```
`apt-get install -y --no-install-recommends`

lthiery replied:
didn't know about this one! thanks

plaes commented on smtc-modem-cores/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+name = "smtc-modem-cores"
+version = "0.1.3"
+authors = ["Louis Thiery <thiery.louis@gmail.com>"]
+edition = "2018"
```
Use 2021 edition? (Here and other `Cargo.toml` files as well.)

lthiery replied:
copy-pasted an old project to get my started! it shows :smile:

plaes commented:
I would rather see smtc-modem-cores (together with -sys) shoul in a separate repository.

lthiery commented:
Rebased on main and restructured: the bindings crates now live in https://github.com/lora-rs/smtc-modem-cores as suggested, so this PR is down to the lora-phy test integration — the smtc crates come in as a git dev-dependency. The set_sleep comparison tests pass against SWL2001 v4.8.0.

lthiery committed:
phy: test sx126x against Semtech's reference driver (SWL2001)

Compare the SPI byte stream our sx126x driver produces against the one
produced by Semtech's reference implementation, via the smtc-modem-cores
bindgen bindings. Starts with set_sleep (cold and warm start).

The bindings crates live in their own repository
(https://github.com/lora-rs/smtc-modem-cores) per review feedback, so
lora-phy only picks them up as a git dev-dependency; no submodule or
in-tree C code here. The bindings crate owns the sx126x_hal_write symbol
and dispatches through a fn-pointer vtable stored in the context, so the
test fixture is plain SpiDevice impls (blocking for the reference
driver, async for ours) recording into a shared op log — no no_mangle
shims or macros on the user side. CI installs cmake/clang to build the C
library for the test and clippy jobs.

lthiery committed:
phy: emulate an sx126x chip to test the full LoRa state machine

A behavioral chip model behind fake SPI/InterfaceVariant impls: writes
execute opcodes against register/buffer/IRQ state, reads answer from it,
and the real Sx126x driver plus LoRa state machine run unmodified on
top. Tests hold a handle to the model to inject received packets and
inspect transmitted ones, covering init, tx, rx (payload + rssi/snr),
and sleep/wake end to end.

This complements the reference-driver byte comparisons: those anchor
what the driver sends, the emulator exercises everything above it.

lthiery committed:
phy: byte-compare the sx126x LoRa config surface against SWL2001

Extends the reference-driver comparisons from set_sleep to standby, rf
frequency (Hz-to-PLL-step math at three frequencies), LoRa modulation
params (including the 500 kHz TxModulation workaround and the LDRO
flag), packet params (including the inverted-IQ workaround), buffer
base address, payload write, set_tx, DIO IRQ masks and IRQ clear.

The fixture now answers reads from canned responses (zeros unprimed),
recorded identically on the blocking and async paths, so read-modify-
write sequences stay comparable. The sync word test primes the chip's
reset values and checks the reference lands on the same register write
our shortcut produces.

lthiery committed:
phy: byte-compare the sx126x RX and CAD paths against SWL2001

do_rx (continuous and single-shot: stop-timer-on-preamble, the LoRa
symbol timeout mantissa/exponent encoding plus its SynchTimeout register
write, rx-boosted gain, SetRx timeout) and do_cad (rx gain, CAD params
per Semtech's application note defaults, SetCAD).

lthiery committed:
phy: byte-compare the sx126x read commands, image calibration and TX CW

Covers get_rx_payload (rx buffer status + read buffer), packet status,
RSSI, IRQ status, ensure_ready's wakeup GetStatus, calibrate_image and
TX continuous wave. The read-path comparisons also check the decoded
values (RSSI/SNR shift math) against the reference's.

The two drivers split read transactions differently — the reference
writes a trailing NOP and reads N bytes, ours reads status plus N — so
fixture equality now compares a wire-canonical form: written bytes with
trailing NOPs trimmed, plus total bytes clocked. MOSI idles at 0x00
during reads, so the chip can't tell the difference.

calibrate_image is compared against cal_img with our table's bytes: the
reference's Hz helper computes ceil(band_end/4), which lands one step
short of datasheet Table 9-2 for most bands (928 MHz -> 0xE8 vs 0xE9).

lthiery committed:
phy: byte-compare sx126x TX power configuration against SWL2001

Covers both PAs across the dBm range: sx1261 (15/14/10/0/-17 plus
clamping from -30) and sx1262 (22/20/17/14/-9 plus clamping from 30,
including the TX clamp workaround read-modify-write), and the slower
init-time ramp.

SWL2001 keeps the power table (datasheet Table 13-21) in the BSP layer
rather than the driver, so the reference is fed the table values our
driver hardcodes; the comparison verifies the SetPaConfig/SetTxParams
encoding around them and pins the table as test data.

lthiery committed:
phy: fix lr1110 command encodings against the SWL2001 reference

Byte-comparison against Semtech's reference driver caught these; the
next commit adds the comparison suite that pins them.

- SetSleep config used the SX126x bit layout (warm start = bit 2); on
  lr11xx warm start is bit 0 and RTC wake-up bit 1, so every sleep was
  a cold sleep losing retention
- SetLoRaSyncTimeout takes the raw symbol count up to 255; the ported
  SX126x mantissa/exponent encoding turned 8 symbols into 32
- WriteBuffer8 has no offset field; the stray 0x00 would have been
  transmitted as the first payload byte
- ReadBuffer8 sent 0 in its length field
- GFSK SetModulationParam takes bitrate in raw bps and deviation in
  raw Hz; the SX126x chip-format conversions programmed garbage
- GetPktStatus (GFSK) response starts with the two RSSI bytes; we read
  the payload length from the rssi_sync byte
- SetGfskSyncWord always carries 8 bytes; shorter words are zero-padded
- AutoTxRx was missing the intermediary mode and timeout fields
- The reference applies the high-ACP workaround (register write, all
  firmware) on every SetTx/SetRx/SetCad; do the same in do_tx/do_rx/
  do_cad so affected firmware (LR1110 0x0303-0x0307, LR1120 0x0101)
  is covered without a manual call

lthiery committed:
phy: byte-compare the lr1110 driver against SWL2001

46 comparison tests mirroring the sx126x suite: system commands
(standby, sleep, reg mode, TCXO, calibration, RF-switch DIO, IRQ
params, version/temp/vbat/random/UID/EUI reads, direct-read status),
the LoRa TX/RX/CAD paths, TX power for all three PAs, regmem, GFSK
configuration and LR-FHSS init/build_frame — plus decoded-value parity
for the read paths.

The lr11xx fixture is simpler than the sx126x one: both drivers split
reads identically (command write, then a discarded status byte plus
data in a second transaction), so transactions compare exactly. Canned
read responses are keyed by the preceding command write; command-less
direct reads consume from a FIFO.

Two documented non-fixes: the reference's GetRandom memcpys the
response into a host-endian u32 where we parse big-endian, and its
generic image-calibration helper lands one step short of the fixed
per-band tables (same story as sx126x cal_img_in_mhz).

Bumps the smtc-modem-cores rev to 42d8559 for the lr11xx bindings.

lthiery committed:
phy: fix sx127x frequency, TX power and RSSI decode against SWL2001

Byte/register comparison against Semtech's reference driver caught
these; the next commit adds the comparison suite that pins them.

- freq_to_pll_step truncated the low 8 pll-step bits (truncate-then-
  shift shortcut), putting fractional-MHz channels like 868.1 and
  903.9 MHz up to ~15 kHz off carrier; now full-precision u64 math
- sx1276 PA_BOOST above 17 dBm: OutputPower used the PaDac-off formula
  (txp - 2) so 18..20 dBm overflowed the 4-bit field — a 20 dBm request
  programmed OutputPower 2 and bled into MaxPower; PaDac-on maps as
  Pout = 5 + OutputPower
- sx1276 RFO at or below 0 dBm: OutputPower went negative and the u8
  cast corrupted the register (PA_BOOST bit included); MaxPower now
  drops to 0 so OutputPower = txp + 4 stays in range, matching the
  reference encoding
- RegDetectionOptimize RMW mask dropped bit 7 (AutomaticIFOn, errata
  2.3 state) on every set_modulation_params
- packet RSSI skipped the 16/15 linearization in the negative-SNR
  branch; the reference and LoRaMac-node apply it in both

lthiery committed:
phy: compare the sx127x driver against SWL2001 at the register level

16 comparison tests. The sx127x is register-based and the two drivers
factor their traffic differently — the reference burst-writes register
runs and keeps shadow copies where ours does single-register RMW over
SPI — so raw transactions can't be compared. Instead the fixture is a
small register file both drivers execute against (writes auto-increment
across bursts, reads answer from the file, FIFO and the write-1-to-
clear RegIrqFlags are kept separate), and equality is the chip-visible
outcome: final register file + FIFO stream. RegOpMode models the
LongRangeMode latch (bit 7 only writable while in sleep and staying in
sleep) since SWL2001 writes mode-only values relying on it.

Covered: standby/sleep transitions, rf freq at 3 channels, LoRa mod
params (SF7/SF12+LDRO/SF6 thresholds), pkt params, sync word, symbol
timeout (both encodings), TX power (PA_BOOST 20/14 dBm, RFO 14/0 dBm),
full TX flow (config -> payload -> IRQ routing -> mode, FIFO stream
byte-compared), CAD flow, RX payload wire traffic, and decoded-value
parity for packet status (SNR exact, RSSI within 1 dB — the two
drivers use different integer approximations of the same 16/15
linearization) and instantaneous RSSI.

Reference findings documented in test comments: SWL2001's
sx127x_set_ocp_value masks the trim with the inverted-form
OCP_TRIM_MASK and so always writes 45 mA; its clear_irq_status and the
whole RX read path are shadow-state only (chip work happens in its DIO
IRQ handlers), so those have no comparable wire footprint.

Bumps smtc-modem-cores to a48bbea (sx127x bindings + vtable; the C
context's radio_id must be set at construction or the driver silently
dispatches to the SX1272 paths).

lthiery committed:
phy: fix the sx1276 500 kHz sensitivity errata frequency gate

The errata 2.1 band edges were written as kHz literals but compared
against frequency_in_hz, so no real channel ever matched and the
RegHighBwOptimize values were never applied — the code always fell
through to the 0x03 reset write. Adds a comparison test pinning the
HF-band values against the ones SWL2001 programs in its own errata
fix.

lthiery committed:
phy: compare against SWL2001 v4.9.0

Bindings rev bump only; the compared functions are line-identical
between v4.8.0 and v4.9.0 and all three suites pass unchanged.

lthiery committed:
phy: apply errata 2.3 (spurious LoRa reception) on the sx1276

The reference driver programs the manual-IF workaround on every SetRx;
ours never did. Applied with the modulation config (the registers only
affect the receiver): non-500 kHz bandwidths clear AutomaticIFOn and
program RegIfFreq1/2 to 0x40/0x00, 500 kHz re-arms AutomaticIFOn.
Bandwidths below 62.5 kHz additionally need the RF frequency shifted up
by one bandwidth, which set_channel owns — those keep chip defaults as
before. Restructures the errata 2.1 block to drop its early returns
(they would have skipped 2.3 exactly when switching back to 500 kHz).
Comparison tests mirror the reference's fix values; a round-trip test
pins the 125->500 kHz AutomaticIFOn re-arm.

lthiery committed:
phy: compare the composed sx126x init_lora sequence end to end

Chains the reference calls (DIO2 RF switch, packet type, sync word,
buffer base, retention list) in init_lora's order and compares the
write stream. The retention-list segment needed prime_read to become a
FIFO per command: both drivers read-modify-write the same retention
register twice in a row, so the second read must see the first write.
Bumps smtc-modem-cores to b31ab4a for the init-path wrappers.

lthiery committed:
phy: grow the sx126x emulator — CAD end to end and RX timeout

The chip model now times out non-continuous receptions when no packet
is pending (the wait collapses to zero — the model has no clock and
nothing can arrive after SetRx executes), which lets the driver's
ReceiveTimeout path run under the real LoRa state machine. New e2e
tests: CAD on quiet and busy channels, and single-mode RX timeout
followed by a successful reception on the same radio.

lthiery commented:
Closing in favor of #445, which carries the same branch history restructured under a clearer description — the scope grew well past the original sx126x set_sleep comparison (all three chips are now covered, with a per-module summary of the 14 driver bugs the comparison caught).
